proposals/0082-relax-transaction-constraints.md

```diff
@@ -0,0 +1,284 @@
+6. The `VersionedMessage` must pass sanitization checks:
+    - The sum of the number of required signatures and the number of read-only
+      unsigned accounts must not be greater than the number of static account
+      keys.
+    - The number of readonly signed accounts must be less than the number of
+      required signatures.
+    - Each lookup table, if present, must be used to load at least one account.
+    - The total number of accounts, static or dynamic, must be less than 256.
+    - Each instruction's `program_id_index` must be less than the number of
+      static account keys.
+    - Each instruction's `program_id_index` must not be the payer index (0).
+    - All account indices in instructions must be less than the number of total
+      accounts.
+7. Transactions that use address lookup tables must be resolvable:
+    - The address lookup table account must exist.
+    - The address lookup table account must be owned by the address lookup
+      table program: `AddressLookupTab1e1111111111111111111111111`
+    - The address lookup table account data must be deserializable into
+      `AddressLookupTable` as defined in `solana-sdk`.
+    - All account table indices specified in the transaction must be less than
+      the number of active addresses in the address lookup table.
+8. Transactions containing pre-compile instructions must pass pre-compile
+   verification checks.
+9. The transaction must not load the same account more than once.
+10. The transaction must have fewer than 64 accounts.
+    - The limit is subject to change to 128 with the activation of
+      `9LZdXeKGeBV6hRLdxS1rHbHoEUsKqesCC2ZAPTPKJAbK`.
+11. The `recent_blockhash` of the transaction message must be valid:
+    - It must exist and not have an age greater than 150.
+    - OR the transaction must be a nonced transaction, and the nonce
+      account must exist and be valid for the given `recent_blockhash`.
+12. The transaction must not have already been processed.
+13. The transaction fee-payer account must:
+    - exist
+    - be owned by the system program: `11111111111111111111111111111111`
+    - have more lamports than the fee
+    - have more lamports than the fee plus the minimum balance
+14. Any non-writable program accounts used in the transaction must exist,
+    regardless of if they are used in any instructions.
+15. The total loaded data size of the transaction must not exceed
+    `requested_loaded_accounts_data_size_limit`, or the default limit (64MB).
+16. Any account used as a program in a top-level instruction must:
+    - be the native loader: `NativeLoader1111111111111111111111111111111`
+    - OR
+      - exist
+      - be executable
+      - be owned by the native loader: `NativeLoader1111111111111111111111111111111`
+    - OR
+      - exist
+      - be executable
+      - the owner account be owned by the native loader: `NativeLoader1111111111111111111111111111111`
+      - the owner account must be executable
+
```

Comment:
We need a complete test suite that tests every single one of these constraints (via txn execution) so that we can also test the behavior when the feature flag is flipped to make sure we relax all the correct constraints

Reply:
Yes, any client should be testing against the protocol, and those tests much change if the protocol changes.

There should already (big should in our codebase...) be tests for all these cases, as they are currently protocol violations, that we must check for.
Tests can be modified to use feature-flag which would change the result of processing pipeline from rejected -> not executed.